enhancements/olm/catalogd-query-endpoint.md

```diff
@@ -0,0 +1,585 @@
+          headers:
+            Last-Modified:
+              schema:
+                format: date-time
+                type: string
+          content:
+            application/jsonl:
+              schema:
+                type: string
+          description: JSONL-formatted catalog FBC
+        "404":
+          content:
+            text/plain:
+              schema:
+                type: string
+          description: ""
+      operationId: getCatalogJSONL
+      description: Retrieve all catalog content
+    parameters:
+      - name: baseURL
+        description: The base API URL from the object's status.urls.base
+        schema:
+          type: string
+        in: path
+        required: true
+```
+<!--
+What are some important details that didn't come across above in the
+**Proposal**? Go in to as much detail as necessary here. This might be
+a good place to talk about core concepts and how they relate. While it is useful
+to go into the details of the code changes required, it is not necessary to show
+how the code will be rewritten in the enhancement.
+-->
+
+### Caching Considerations
+
+This proposal indends to provide RFC7234 caching compliance through support for `Last-Modified` and `If-Modified-Since` directives. Clients can use these headers to avoid re-downloading unchanged data. 
+The server will respond with 304 Not Changed if the catalog metadata is unchanged.  
+                  
+Clients are also encouraged to implement local caching for frequently queried metadata.
+
+
+### Risks and Mitigations
+
+Depending on the implementation of this proposal, exercise of this endpoint could function as an I/O multiplier which could starve other clients which are dependent on the same API. 
+The first iteration of this endpoint will not attempt to mitigate denial-of-service attempts via rate-limiting or other resource consumption constraints.
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside your immediate sub-project.
+-->
+
+### Drawbacks
+
+Given the highly-variable nature of responses based on unbound combinations of schema/name/package, any server-side caching will be problematic if not completely impractical. Read: computationally and/or storage expensive.
+
+For example, response for `query?schema=olm.package&name=foo` and `query?schema=olm.package&name=bar` represent disjoint sets, and the server would have to either ignore `If-Modified-Since` directives in client requests or rely on their veracity uncritically. 
+
+Further, if the underlying implementation relies on anything other than a dictionary of cached requests (&parameters) to dictate date-relevant responses (for e.g. relying on the catalog unpack time), then any server-side caching would be erroneous.  
+
```

Comment:
Other drawbacks:
- Server implementation is much more complex and resource intensive due to the need for indexing and query handling
- The "all" endpoint incentivizes being a good steward of resources. It likely pushes clients toward an architecture of downloading once and using the local cache rather than making many (potentially duplicate) requests to the server (potentially without caching)
- The "all" endpoint returns valid FBC. This new input generally won't. Will clients make incorrect representations of the catalog due to misunderstandings of what constitutes valid FBC?